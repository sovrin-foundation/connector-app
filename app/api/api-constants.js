// @flow

export const TOKEN_EXPIRED_CODE = 'VES-101'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_REQUEST_ID = 'id'
export const QR_CODE_SENDER_DETAIL = 's'
export const QR_CODE_SENDER_DID = 'd'
export const QR_CODE_SENDER_NAME = 'n'
export const QR_CODE_LOGO_URL = 'l'
export const QR_CODE_SENDER_VERIFICATION_KEY = 'v'
export const QR_CODE_SENDER_KEY_DELEGATION = 'dp'
export const QR_CODE_DELEGATION_DID = 'd'
export const QR_CODE_DELEGATION_KEY = 'k'
export const QR_CODE_DELEGATION_SIGNATURE = 's'
export const QR_CODE_SENDER_AGENCY = 'sa'
export const QR_CODE_SENDER_AGENCY_DID = 'd'
export const QR_CODE_SENDER_AGENCY_KEY = 'v'
export const QR_CODE_SENDER_AGENCY_ENDPOINT = 'e'
export const QR_CODE_TARGET_NAME = 't'
export const GENESIS_FILE_NAME = 'pool_transactions_genesis'

export const MESSAGE_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'credOffer',
  CLAIM_REQUEST: 'credReq',
  PROOF_REQUEST: 'proofReq',
  PROOF: 'proof',
  CLAIM: 'cred',
}

export const PAYLOAD_TYPE = {
  GET_MESSAGE: 'GET_MSGS',
  UPDATE_PUSH_COM_METHOD: 'UPDATE_PUSH_COM_METHOD',
  SEND_MESSAGE: 'SEND_MSG',
}

export const STATUS_CODE = {
  ACCEPT: 'MS-104',
}

export const ERROR_ALREADY_EXIST = {
  code: 'OCS',
  message: 'You already have a relationship with ',
}

export const ERROR_ALREADY_EXIST_TITLE = 'Relationship already established'

export const ERROR_INVITATION_RESPONSE_PARSE_CODE = 'RN-100'

export const ERROR_INVITATION_RESPONSE_PARSE =
  'Error parsing error message in invitation response'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE = 'RN-101'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE =
  'Error parsing error message Pending Invitation SMS'

export const ERROR_INVITATION_RESPONSE_FAILED =
  'Failed to establish connection. Please try again later.'
