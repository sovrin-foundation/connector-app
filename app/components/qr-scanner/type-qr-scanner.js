// @flow
import type { InvitationPayload } from '../../invitation/type-invitation'

export const SCAN_STATUS = {
  SCANNING: 'scanning...',
  SUCCESS: 'Success!',
  FAIL: 'QR code format is invalid.',
  DOWNLOADING_INVITATION: 'Downloading invitation...',
  NO_INVITATION_DATA: 'No message found behind this QR code.',
}

export type QrCode = {
  id: string,
  s: {
    n: string,
    dp: {
      d: string,
      k: string,
      s: string,
    },
    d: string,
    l: string,
    v: string,
  },
  sa: {
    d: string,
    v: string,
    e: string,
  },
  t: string,
}

type ValuesType = <V>(v: V) => V

type QR_SCAN_STATUS =
  | typeof SCAN_STATUS.SCANNING
  | typeof SCAN_STATUS.FAIL
  | typeof SCAN_STATUS.SUCCESS
  | typeof SCAN_STATUS.DOWNLOADING_INVITATION
  | typeof SCAN_STATUS.NO_INVITATION_DATA

export type QrScannerState = {
  scanning: boolean,
  scanStatus: QR_SCAN_STATUS,
  cameraActive?: boolean,
}

export type QrScannerProps = {
  onRead: QrCode => void,
  onClose: () => void,
  onEnvironmentSwitchUrl: EnvironmentSwitchUrlQrCode => void,
  onInvitationUrl: InvitationPayload => void,
}

export type CameraMarkerProps = {
  status: QR_SCAN_STATUS,
  onClose: () => void,
}

export const TOP_LEFT = 'topLeft'
export const TOP_RIGHT = 'topRight'
export const BOTTOM_LEFT = 'bottomLeft'
export const BOTTOM_RIGHT = 'bottomRight'

export type CornerBoxProps = {
  status: QR_SCAN_STATUS,
  position:
    | typeof TOP_LEFT
    | typeof TOP_RIGHT
    | typeof BOTTOM_LEFT
    | typeof BOTTOM_RIGHT,
}

export type EnvironmentSwitchUrlQrCode = {
  url: string,
  name: string,
}

export type InvitationUrlQrCode = {
  url: string,
}
