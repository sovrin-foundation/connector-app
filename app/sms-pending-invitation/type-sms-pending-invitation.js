// @flow
import type {
  CustomError,
  InitialTestAction,
  ResetAction,
} from '../common/type-common'

export const SMSPendingInvitationStatus = {
  NONE: 'NONE',
  FETCH_FAILED: 'FETCH_FAILED',
  RECEIVED: 'RECEIVED',
  SEEN: 'SEEN',
}

export type SMSPendingInvitationStatusType = $Keys<
  typeof SMSPendingInvitationStatus
>

export type AgentKeyDelegationProof = {
  agentDID: string,
  agentDelegatedKey: string,
  signature: string,
}

export type InvitationSenderDetail = {
  name: string,
  agentKeyDlgProof: AgentKeyDelegationProof,
  DID: string,
  logoUrl: string,
  verKey: string,
}

export type InvitationSenderAgencyDetail = {
  DID: string,
  verKey: string,
  endpoint: string,
}

export type SMSPendingInvitationPayload = {
  connReqId: string,
  statusCode: string,
  senderDetail: InvitationSenderDetail,
  senderAgencyDetail: InvitationSenderAgencyDetail,
  targetName: string,
}

export type SMSPendingInvitation = {
  +payload: ?SMSPendingInvitationPayload,
  +status: SMSPendingInvitationStatusType,
  +isFetching: boolean,
  +error?: ?CustomError,
}

export type SMSPendingInvitations = Array<
  SMSPendingInvitation & { invitationToken: string }
>

export type SMSPendingInvitationStore = {
  +[string]: SMSPendingInvitation,
}

export const SMS_PENDING_INVITATION_REQUEST: 'SMS_PENDING_INVITATION_REQUEST' =
  'SMS_PENDING_INVITATION_REQUEST'

export type SMSPendingInvitationRequestAction = {
  type: typeof SMS_PENDING_INVITATION_REQUEST,
  smsToken: string,
}

export const SMS_PENDING_INVITATION_RECEIVED: 'SMS_PENDING_INVITATION_RECEIVED' =
  'SMS_PENDING_INVITATION_RECEIVED'

export type SMSPendingInvitationReceivedAction = {
  type: typeof SMS_PENDING_INVITATION_RECEIVED,
  data: SMSPendingInvitationPayload,
  smsToken: string,
}

export const SMS_PENDING_INVITATION_FAIL: 'SMS_PENDING_INVITATION_FAIL' =
  'SMS_PENDING_INVITATION_FAIL'

export type SMSPendingInvitationFailAction = {
  type: typeof SMS_PENDING_INVITATION_FAIL,
  error: CustomError,
  smsToken: string,
}

export const SMS_PENDING_INVITATION_SEEN: 'SMS_PENDING_INVITATION_SEEN' =
  'SMS_PENDING_INVITATION_SEEN'

export type SMSPendingInvitationSeenAction = {
  type: typeof SMS_PENDING_INVITATION_SEEN,
  smsToken: string,
}

export const SAFE_TO_DOWNLOAD_SMS_INVITATION = 'SAFE_TO_DOWNLOAD_SMS_INVITATION'

export type SafeToDownloadSmsInvitationAction = {
  type: typeof SAFE_TO_DOWNLOAD_SMS_INVITATION,
}

export type SMSPendingInvitationAction =
  | SMSPendingInvitationRequestAction
  | SMSPendingInvitationReceivedAction
  | SMSPendingInvitationFailAction
  | SMSPendingInvitationSeenAction
  | SafeToDownloadSmsInvitationAction
  | InitialTestAction
  | ResetAction
