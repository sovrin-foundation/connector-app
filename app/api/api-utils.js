// @flow
import { SERVER_ERROR_CODE, SERVER_API_CALL_ERROR } from './api-constants'
import { captureError } from '../services/error/error-handler'
import type { CustomError } from '../common/type-common'
import type { ApiData, BackendError } from './type-api'

export const options = (
  method: string = 'GET',
  body: ?{ [string]: any } = null
) => {
  let data: ApiData = {
    method,
    mode: 'cors',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  }

  if (body) {
    data.body = JSON.stringify(body)
  }

  return data
}

export const api = (
  url: string,
  apiOptions: ApiData,
  showAlert: boolean = false
) =>
  fetch(url, apiOptions)
    .then(res => {
      // TODO:KS create common method to return successful
      // and unsuccessful api response
      if (res.status >= 200 && res.status < 300) {
        return res
          .json()
          .then(response => ({
            payload: response,
          }))
          .catch((error: Error) => ({
            // our api call was successful,
            // however server did not return any json response
            // Ideally server should send such response with status code 204
            // so that we don't try to parse response for such methods.
            // As of now just send an empty payload signalling that call went fine
            payload: {},
          }))
      } else {
        // Fail with error code if status code is above 300
        return res.text().then((response: string) => ({
          error: response,
        }))
      }
    })
    .then(response => {
      // if response contains payload and no error, that means we got success
      if (response.payload && !response.error) {
        return response.payload
      } else {
        let errorResponse: CustomError = {
          message: response.error,
          code: SERVER_ERROR_CODE,
        }

        try {
          // try to convert error response to json, if it fails that means
          // we did not get error code and message
          const backendError = (JSON.parse(response.error): BackendError)
          errorResponse = {
            message: backendError.statusMsg,
            code: backendError.statusCode,
          }
        } catch (e) {
          // since we did not get error code and message
          // let's just use default that we assigned above,
          // we don't need to do anything here
        }

        throw new Error(JSON.stringify(errorResponse))
      }
    })
    // .catch(captureError) Commenting this for now, it is not returning errors back to saga as of now, Fix this
    .catch((e: Error) => {
      // we don't want to throw JavaScript with simple message
      // we want it to be JSON parse compatible, so that error handlers
      // can parse and use this error
      // Here we are handling network and other http status code errors
      let error = {
        message: e.message,
        code: SERVER_API_CALL_ERROR,
      }

      try {
        // if we already converted JavaScript error to CustomError
        // then pass through same error
        error = JSON.parse(e.message)
      } finally {
      }

      throw new Error(JSON.stringify(error))
    })
