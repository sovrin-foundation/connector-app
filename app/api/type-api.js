// @flow
import type { GenericObject } from '../common/type-common'
import { MESSAGE_TYPE } from './api-constants'
import type { IndyClaimRequest } from '../bridge/react-native-cxs/type-cxs'
import type { ProofApiData } from '../proof-request/type-proof-request'

export type ApiData = {
  method: string,
  mode: string,
  headers: { [string]: string },
  body?: string,
}

export type BackendError = {
  statusCode: string,
  statusMsg: string,
}

export type GetInvitationLinkApiData = {
  agencyUrl: string,
  smsToken: string,
}

export type MessageApiData = {
  agencyUrl: string,
  dataBody: GenericObject,
}

export type DownloadInvitationApiData = {
  url: string,
}

export type SendAuthenticationResponseApiData = {
  data: {
    identifier: string,
    dataBody: GenericObject,
  },
  config: {
    agencyUrl: string,
  },
}

export type SendMessageApiData = {
  agencyUrl: string,
  userPairwiseDid: string,
  agentPayload: GenericObject,
}

export type ApiClaimRequest = {
  userPairwiseDid: string,
  remoteDid: string,
} & IndyClaimRequest

export type SendClaimRequestApiData = {
  agencyUrl: string,
  userPairwiseDid: string,
  responseMsgId: string,
  claimRequest: ApiClaimRequest,
}
export type SendProofApiData = {
  agencyUrl: string,
  userPairwiseDid: string,
  responseMsgId: string,
  proof: ProofApiData,
}

export type EdgeClaimRequest = {
  msg_type: typeof MESSAGE_TYPE.CLAIM_REQUEST,
  version: string,
  to_did: string,
  from_did: string,
  tid: string,
  mid: string,
} & IndyClaimRequest

export type EnvironmentDetailUrlDownloaded = {
  agencyDID: string,
  agencyUrl: string,
  agencyVerificationKey: string,
  poolConfig: string,
}
