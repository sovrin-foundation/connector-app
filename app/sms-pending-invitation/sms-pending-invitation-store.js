// @flow
import {
  put,
  takeLatest,
  takeEvery,
  call,
  take,
  all,
  select,
} from 'redux-saga/effects'
import type { CustomError } from '../common/type-common'
import type {
  SMSPendingInvitationStore,
  SMSPendingInvitationAction,
  SMSPendingInvitationStatusType,
  SMSPendingInvitationPayload,
  SMSPendingInvitationRequestAction,
} from './type-sms-pending-invitation'
import type { InvitationPayload } from '../invitation/type-invitation'
import {
  SMS_PENDING_INVITATION_REQUEST,
  SMS_PENDING_INVITATION_RECEIVED,
  SMS_PENDING_INVITATION_FAIL,
  SMS_PENDING_INVITATION_SEEN,
  SMSPendingInvitationStatus,
  SAFE_TO_DOWNLOAD_SMS_INVITATION,
} from './type-sms-pending-invitation'
import { invitationDetailsRequest, getInvitationLink } from '../api/api'
import {
  ERROR_PENDING_INVITATION_RESPONSE_PARSE,
  ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE,
} from '../api/api-constants'
import {
  getAgencyUrl,
  getCurrentScreen,
  getHydrationState,
} from '../store/store-selector'
import { invitationReceived } from '../invitation/invitation-store'
import {
  HYDRATED,
  UNSAFE_SCREENS_TO_DOWNLOAD_SMS,
} from '../store/type-config-store'
import { RESET } from '../common/type-common'
import { captureError } from '../services/error/error-handler'
import { isValidInvitationUrl } from '../components/qr-scanner/qr-invitation-url-validator'

const initialState = {}

export const getSmsPendingInvitation = (smsToken: string) => ({
  type: SMS_PENDING_INVITATION_REQUEST,
  smsToken,
})

export const smsPendingInvitationReceived = (
  smsToken: string,
  data: SMSPendingInvitationPayload
) => ({
  type: SMS_PENDING_INVITATION_RECEIVED,
  data,
  smsToken,
})

export const smsPendingInvitationFail = (
  smsToken: string,
  error: CustomError
) => ({
  type: SMS_PENDING_INVITATION_FAIL,
  smsToken,
  error,
})

export const smsPendingInvitationSeen = (smsToken: string) => ({
  type: SMS_PENDING_INVITATION_SEEN,
  smsToken,
})

// Below action tells that now user can't change the environment
// and so whatever environment we have available now, we can go ahead and
// start downloading SMS invitation from that environment
export const safeToDownloadSmsInvitation = () => ({
  type: SAFE_TO_DOWNLOAD_SMS_INVITATION,
})

export const convertSmsPayloadToInvitation = (
  pendingInvitation: SMSPendingInvitationPayload
): InvitationPayload => ({
  senderEndpoint: pendingInvitation.senderAgencyDetail.endpoint,
  requestId: pendingInvitation.connReqId,
  senderAgentKeyDelegationProof:
    pendingInvitation.senderDetail.agentKeyDlgProof,
  senderName: pendingInvitation.senderDetail.name,
  senderDID: pendingInvitation.senderDetail.DID,
  senderLogoUrl: pendingInvitation.senderDetail.logoUrl,
  senderVerificationKey: pendingInvitation.senderDetail.verKey,
  targetName: pendingInvitation.targetName,
  senderAgencyDetail: pendingInvitation.senderAgencyDetail,
  senderDetail: pendingInvitation.senderDetail,
})

export function* callSmsPendingInvitationRequest(
  action: SMSPendingInvitationRequestAction
): Generator<*, *, *> {
  // we wait to download SMS invitation till we know that we are safe
  // because user can still change the environment and point to something else
  // to download sms invitation from updated environment
  // if a user is gone past lock selection screen, then we know we are safe
  // or user has already setup lock and is now coming for second SMS invitation
  // or user is not on one of these screens
  // splash screen, lock selection, dev environment switch
  const currentScreen: string = yield select(getCurrentScreen)
  if (UNSAFE_SCREENS_TO_DOWNLOAD_SMS.indexOf(currentScreen) > -1) {
    // user is on screens where he has chance to change environment details
    // so we wait for event which tells that we are safe
    yield take(SAFE_TO_DOWNLOAD_SMS_INVITATION)
  }

  // if we reach this point, that means either user was not on some unsafe screen
  // or we waited till safe action was raised, either way
  // we can now go ahead and start downloading SMS invitation

  // I lied a bit in above statement that we can start downloading SMS now
  // there is one more scenario that we need to consider before we can start
  // downloading sms, this saga can also be triggered if user already has an app
  // and user switched environments while setting up the app first time
  // so now we need to maintain switched environment for the lifetime of app
  // so we store those environment details in phone storage and
  // hydrate them when app is killed and started again
  // now, when user starts the app second time, user's updated config
  // might still not be fetched or app is not fully hydrated yet,
  // so, we wait for app to be hydrated if not already hydrated
  const isAppHydrated: boolean = yield select(getHydrationState)
  if (!isAppHydrated) {
    yield take(HYDRATED)
  }

  // now we are sure that user can't change the environment for first time app
  // and also we are sure that if user did update the environment first time
  // then we have got the updated config
  // and now, we can go ahead and start downloading SMS pending invitation

  const agencyUrl: string = yield select(getAgencyUrl)
  const { smsToken } = action

  try {
    // there can be two types of token that we can get from deep link
    // 1. it can be 8 characters token
    // 2. it can be a url that would return invitation details

    if (!smsToken) {
      throw new Error('No data in token received')
    }

    let invitationLink = {
      url: '',
    }
    if (smsToken.length < 12) {
      // we are assuming that if length is less than 12, then we got a token
      // get invitation link
      invitationLink = yield call(getInvitationLink, {
        agencyUrl,
        smsToken,
      })
    } else if (isValidInvitationUrl(smsToken)) {
      // check if we got the url, then directly assign url
      invitationLink = {
        url: smsToken,
      }
    } else {
      throw new Error('Not recognized for any type of token that we know')
    }

    // get pending invitation
    // TODO:KS Handle error for this API, all business logic and API related
    const pendingInvitationPayload = yield call(invitationDetailsRequest, {
      url: invitationLink.url,
    })

    yield put(
      invitationReceived({
        payload: convertSmsPayloadToInvitation(pendingInvitationPayload),
      })
    )
    yield put(smsPendingInvitationReceived(smsToken, pendingInvitationPayload))
  } catch (e) {
    let error: CustomError = {
      code: ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE,
      message: `${ERROR_PENDING_INVITATION_RESPONSE_PARSE}: ${e.message}`,
    }

    try {
      error = JSON.parse(e.message)
    } catch (e) {}
    captureError(e)
    yield put(smsPendingInvitationFail(smsToken, error))
  }
}

function* watchSmsPendingInvitationRequest(): any {
  yield takeEvery(
    SMS_PENDING_INVITATION_REQUEST,
    callSmsPendingInvitationRequest
  )
}

export function* watchSmsPendingInvitationSaga(): any {
  yield all([watchSmsPendingInvitationRequest()])
}

export default function smsPendingInvitationReducer(
  state: SMSPendingInvitationStore = initialState,
  action: SMSPendingInvitationAction
) {
  switch (action.type) {
    case SMS_PENDING_INVITATION_REQUEST:
      if (state[action.smsToken] !== undefined) {
        return state
      } else {
        return {
          ...state,
          [action.smsToken]: {
            payload: null,
            isFetching: true,
            status: SMSPendingInvitationStatus.NONE,
            error: null,
          },
        }
      }

    case SMS_PENDING_INVITATION_RECEIVED:
      return {
        ...state,
        [action.smsToken]: {
          ...state[action.smsToken],
          isFetching: false,
          payload: action.data,
          status: SMSPendingInvitationStatus.RECEIVED,
        },
      }

    case SMS_PENDING_INVITATION_FAIL:
      return {
        ...state,
        [action.smsToken]: {
          ...state[action.smsToken],
          payload: null,
          isFetching: false,
          error: action.error,
          status: SMSPendingInvitationStatus.FETCH_FAILED,
        },
      }

    case SMS_PENDING_INVITATION_SEEN:
      if (action.smsToken) {
        return {
          ...state,
          [action.smsToken]: {
            ...state[action.smsToken],
            status: SMSPendingInvitationStatus.SEEN,
          },
        }
      } else {
        return state
      }

    case RESET:
      return initialState

    default:
      return state
  }
}
