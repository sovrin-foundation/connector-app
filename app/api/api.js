// @flow
import { api, options } from './api-utils'
import type {
  GetInvitationLinkApiData,
  MessageApiData,
  DownloadInvitationApiData,
  SendAuthenticationResponseApiData,
  EnvironmentDetailUrlDownloaded,
} from './type-api'
import { PAYLOAD_TYPE, MESSAGE_TYPE, STATUS_CODE } from './api-constants'

export const getInvitationLink = ({
  agencyUrl,
  smsToken,
}: GetInvitationLinkApiData) =>
  api(`${agencyUrl}/agency/url-mapper/${smsToken}`, options('GET'))

export const invitationDetailsRequest = ({ url }: DownloadInvitationApiData) =>
  api(url, options('GET'))

export const getAdditionalData = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendAuthenticationRequest = ({
  data: { identifier, dataBody },
  config: { agencyUrl },
}: SendAuthenticationResponseApiData) =>
  api(`${agencyUrl}/agent/${identifier}/auth`, options('PUT', dataBody), true)

export const downloadEnvironmentDetails = (
  url: string
): Promise<EnvironmentDetailUrlDownloaded> => api(url, options())
