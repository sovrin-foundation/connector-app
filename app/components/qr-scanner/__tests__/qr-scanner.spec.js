// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { SCAN_STATUS } from '../type-qr-scanner'
import QRScanner from '../qr-scanner'
import {
  qrData,
  validQrCodeEnvironmentSwitchUrl,
  validInvitationUrlQrCode,
  smsDownloadedPayload,
} from '../../../../__mocks__/static-data'
import * as api from '../../../api/api'
import { convertSmsPayloadToInvitation } from '../../../sms-pending-invitation/sms-pending-invitation-store'

describe('<QRScanner />', () => {
  const getProps = () => ({
    onClose: jest.fn(),
    onRead: jest.fn(),
    onEnvironmentSwitchUrl: jest.fn(),
    onInvitationUrl: jest.fn(),
  })

  function setup() {
    const props = getProps()
    const wrapper = renderer.create(<QRScanner {...props} />)
    const instance = wrapper.getInstance()
    return {
      ...props,
      instance,
      wrapper,
    }
  }

  it('should match snapshot', () => {
    const { wrapper } = setup()
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should call onRead once QR code read is successful', () => {
    jest.useFakeTimers()
    const { onRead, instance } = setup()

    const qrReadEvent = {
      data: JSON.stringify(qrData),
    }

    instance.onRead(qrReadEvent)
    expect(onRead).toHaveBeenCalledWith(expect.objectContaining(qrData))
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SUCCESS)
  })

  it('should set state to fail if QR code is not correct', () => {
    jest.useFakeTimers()

    const { instance } = setup()

    instance.onRead({ data: '' })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.FAIL)

    jest.runAllTimers()

    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })

  it('should call onEnvironmentSwitchUrl if it reads correct environment switcher url', () => {
    jest.useFakeTimers()
    const { onEnvironmentSwitchUrl, instance } = setup()

    instance.onRead({ data: validQrCodeEnvironmentSwitchUrl })
    expect(onEnvironmentSwitchUrl).toHaveBeenCalledWith({
      name: 'dev',
      url: validQrCodeEnvironmentSwitchUrl,
    })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SUCCESS)
  })

  it('should send a request to download invitation if url is scanned', async () => {
    jest.useFakeTimers()
    const invitationDetailRequestSpy = jest.spyOn(
      api,
      'invitationDetailsRequest'
    )

    invitationDetailRequestSpy.mockImplementation(() =>
      Promise.resolve(smsDownloadedPayload)
    )

    const { onInvitationUrl, instance } = setup()

    const pendingQrProcessing = instance.onRead({
      data: validInvitationUrlQrCode,
    })
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.DOWNLOADING_INVITATION)
    // process API call
    await pendingQrProcessing

    expect(onInvitationUrl).toHaveBeenCalledWith(
      convertSmsPayloadToInvitation(smsDownloadedPayload)
    )
    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SUCCESS)

    invitationDetailRequestSpy.mockReset()
    invitationDetailRequestSpy.mockRestore()

    jest.runAllTimers()

    expect(instance.state.scanStatus).toBe(SCAN_STATUS.SCANNING)
  })
})
