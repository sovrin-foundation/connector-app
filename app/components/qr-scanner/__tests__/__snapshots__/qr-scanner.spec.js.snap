// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QRScanner /> should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 20,
            },
            Object {
              "fontFamily": "Lato",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            Object {
              "fontWeight": "600",
            },
            null,
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
      >
        Scan QR Code
      </Text>
      <View
        spaceBetween={true}
        style={
          Object {
            "height": 250,
            "justifyContent": "space-between",
            "marginVertical": 50,
            "width": 250,
          }
        }
        transparentBg={true}
      >
        <View
          row={true}
          spaceBetween={true}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              Object {
                "borderColor": "#ffffff",
                "borderLeftWidth": 5,
                "borderTopWidth": 5,
                "height": 70,
                "width": 70,
              }
            }
            transparentBg={true}
          />
          <View
            style={
              Object {
                "borderColor": "#ffffff",
                "borderRightWidth": 5,
                "borderTopWidth": 5,
                "height": 70,
                "width": 70,
              }
            }
            transparentBg={true}
          />
        </View>
        <View
          row={true}
          spaceBetween={true}
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              Object {
                "borderBottomWidth": 5,
                "borderColor": "#ffffff",
                "borderLeftWidth": 5,
                "height": 70,
                "width": 70,
              }
            }
            transparentBg={true}
          />
          <View
            style={
              Object {
                "borderBottomWidth": 5,
                "borderColor": "#ffffff",
                "borderRightWidth": 5,
                "height": 70,
                "width": 70,
              }
            }
            transparentBg={true}
          />
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.2)",
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": -1,
          }
        }
      />
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "Lato",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            Object {
              "fontWeight": "600",
            },
            null,
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "color": "#ffffff",
            },
            Object {
              "marginVertical": 30,
            },
          ]
        }
      >
        scanning...
      </Text>
      <View
        accessibilityLabel="close-qr-scanner-container"
        accessible={true}
        center={true}
        row={true}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginVertical": 20,
          }
        }
        testID="close-qr-scanner-container"
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel="close-qr-scanner-icon-touchable"
          accessibilityTraits={undefined}
          accessible={false}
          clearBg={true}
          hitSlop={
            Object {
              "bottom": 30,
              "left": 30,
              "right": 30,
              "top": 30,
            }
          }
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          small={true}
          src={1}
          style={
            Object {
              "backgroundColor": "transparent",
            }
          }
          testID="close-qr-scanner-icon-touchable"
        >
          <View
            style={Object {}}
          >
            <Image
              accessibilityLabel="close-qr-scanner-icon"
              accessible={true}
              resizeMode="contain"
              source={1}
              style={
                Array [
                  Object {
                    "height": 20,
                    "width": 20,
                  },
                  null,
                ]
              }
              testID="close-qr-scanner-icon"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
